src/jit/gentree.cpp

```diff
@@ -15104,157 +15104,121 @@ GenTree* Compiler::gtBuildCommaList(GenTree* list, GenTree* expr)
-    {
-        // Add the side effect to the list and return
-        //
-        *pList = gtBuildCommaList(*pList, expr);
-        return;
-    }
-
-    if (kind & GTK_LEAF)
-    {
-        return;
-    }
-
-    if (oper == GT_LOCKADD || oper == GT_XADD || oper == GT_XCHG || oper == GT_CMPXCHG)
-    {
-        // These operations are kind of important to keep
-        *pList = gtBuildCommaList(*pList, expr);
-        return;
-    }
-
-    if (kind & GTK_SMPOP)
+    class SideEffectExtractor final : public GenTreeVisitor<SideEffectExtractor>
     {
-        GenTree* op1 = expr->gtOp.gtOp1;
-        GenTree* op2 = expr->gtGetOp2IfPresent();
+    public:
+        const unsigned       m_flags;
+        ArrayStack<GenTree*> m_sideEffects;
 
-        if (flags & GTF_EXCEPT)
+        enum
         {
-            // Special case - GT_ADDR of GT_IND nodes of TYP_STRUCT
-            // have to be kept together
-
-            if (oper == GT_ADDR && op1->OperIsIndir() && op1->gtType == TYP_STRUCT)
-            {
-                *pList = gtBuildCommaList(*pList, expr);
+            DoPreOrder        = true,
+            UseExecutionOrder = true
+        };
 
-#ifdef DEBUG
-                if (verbose)
-                {
-                    printf("Keep the GT_ADDR and GT_IND together:\n");
-                }
-#endif
-                return;
-            }
+        SideEffectExtractor(Compiler* compiler, unsigned flags)
+            : GenTreeVisitor(compiler), m_flags(flags), m_sideEffects(compiler->getAllocator(CMK_SideEffects))
+        {
         }
 
-        /* Continue searching for side effects in the subtrees of the expression
-         * NOTE: Be careful to preserve the right ordering - side effects are prepended
-         * to the list */
-
-        /* Continue searching for side effects in the subtrees of the expression
-         * NOTE: Be careful to preserve the right ordering
-         * as side effects are prepended to the list */
-
-        if (expr->gtFlags & GTF_REVERSE_OPS)
+        fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
         {
-            assert(oper != GT_COMMA);
-            if (op1)
+            GenTree* node = *use;
+
+            if (!m_compiler->gtTreeHasSideEffects(node, m_flags))
             {
-                gtExtractSideEffList(op1, pList, flags);
+                return Compiler::WALK_SKIP_SUBTREES;
             }
-            if (op2)
+
+            if (m_compiler->gtNodeHasSideEffects(node, m_flags))
             {
-                gtExtractSideEffList(op2, pList, flags);
+                m_sideEffects.Push(node);
+                return Compiler::WALK_SKIP_SUBTREES;
             }
-        }
-        else
-        {
-            if (op2)
+
+            // These have GTF_ASG but for some reason gtNodeHasSideEffects ignores them.
+            // Anyway, they must be preserved, no matter what side effect flags are passed in.
+            if (node->OperIs(GT_LOCKADD, GT_XADD, GT_XCHG, GT_CMPXCHG))
```

Comment:
I believe that thee four shoudl be considered as assignments as they do in fact write to memory (atomically):

GT_LOCKADD, GT_XADD, GT_XCHG, GT_CMPXCHG

MemoryBarrier doesn't write to memory so it might not need to be mark with an assignment, but it does need to always be preserved.  It may simplify the implementation if it is also considered an atomic memory operations like the others.

Reply:
I've made a compromise and replaced `OperIs(GT_LOCKADD...)` with `OperIsAtomicOp()` since it already exists and it's used in a number of places. At the same time, I've added TODO-Cleanup comments about the whole `OperRequiresAsgFlag` situation.

> MemoryBarrier doesn't write to memory so it might not need to be mark with an assignment, but it does need to always be preserved. It may simplify the implementation if it is also considered an atomic memory operations like the others.

Yep. And I wonder if we shouldn't also treat volatile indirs as assignment side effects as well, they're after all half memory barriers and currently they are dropped - #18748